File: demo_engine.py
"""cga 渲染引擎 demo: three.js 风格场景 + 轨道动画 → PNG 帧 + GIF。

运行: uv run python demo_engine.py [帧数] [输出目录]
输出: <out>/frame_%03d.png + <out>/orbit.gif

场景: 地面 + 红/蓝球 + 金柱 + 绿盒 + 紫圆盘, 平行光 + 点光 + 环境光。
"""

import sys
from pathlib import Path

import numpy as np
from PIL import Image

from cga.engine import (
    AmbientLight,
    BoxGeometry,
    CircleGeometry,
    Color,
    CylinderGeometry,
    DirectionalLight,
    Mesh,
    MeshStandardMaterial,
    OrbitControls,
    PerspectiveCamera,
    PlaneGeometry,
    PointLight,
    Renderer,
    Scene,
    SphereGeometry,
)


def build_scene() -> Scene:
    scene = Scene()
    scene.add(
        Mesh(
            PlaneGeometry((0, 1, 0), 0.0),
            # 半光泽地面: roughness 0.7 → Blinn 指数 ~22, 掠射高光聚焦不整片过曝
            MeshStandardMaterial(Color(0xB0B0B0), roughness=0.7),
        ),
        # v1 无 envMap/IBL: 高 metalness 会黑死 (three.js 同样问题)。
        # 压低 metalness, 保留金属色高光, 让漫反射色可读。
        Mesh(
            SphereGeometry(1.0),
            MeshStandardMaterial(Color(0xC0392B), roughness=0.25, metalness=0.25),
            position=(0, 1, 0),
        ),
        Mesh(
            SphereGeometry(0.6),
            MeshStandardMaterial(Color(0x2980B9), roughness=0.15, metalness=0.35),
            position=(-2.2, 0.6, 0.5),
        ),
        Mesh(
            CylinderGeometry(0.7),
            MeshStandardMaterial(Color(0xD4AC0D), roughness=0.4, metalness=0.3),
            position=(2.2, 0.7, -0.5),
        ),
        Mesh(
            BoxGeometry(0.9, 0.9, 0.9),
            MeshStandardMaterial(Color(0x27AE60), roughness=0.6),
            position=(0.8, 0.45, 1.8),
        ),
        Mesh(
            CircleGeometry(0.9),
            MeshStandardMaterial(Color(0x8E44AD), roughness=0.3),
            position=(-2.4, 2.2, 0.8),
            rotation_axis=(1, 0, 0),
            rotation_angle=-0.4,
        ),
        DirectionalLight(intensity=0.38, direction=(0.4, 1.0, 0.35)),  # 主光
        PointLight(intensity=0.7, position=(0, 4, 3.5)),  # 顶面点缀
        DirectionalLight(intensity=0.18, direction=(0, 0.35, 0.9)),  # 正面补光
        AmbientLight(intensity=0.34),  # 环境底光: 金属暗面可读
    )
    return scene


def main() -> None:
    frames = int(sys.argv[1]) if len(sys.argv) > 1 else 90
    out_dir = Path(sys.argv[2] if len(sys.argv) > 2 else "artifacts")
    out_dir.mkdir(exist_ok=True)

    scene = build_scene()
    camera = PerspectiveCamera(
        fov=50, aspect=4 / 3, position=(0, 2.4, 6.2), target=(0, 0.8, 0)
    )
    camera.look_at((0, 0.8, 0))
    controls = OrbitControls(camera, target=(0, 0.8, 0), radius=6.6, elevation=0.42)
    renderer = Renderer(360, 270)

    paths: list[Path] = []
    for i in range(frames):
        controls.azimuth = 2.0 * 3.141592653589793 * i / frames  # 绕一圈
        controls.elevation = 0.42 + 0.12 * np.sin(4.0 * 3.141592653589793 * i / frames)
        controls.update()
        img = renderer.render(scene, camera)
        arr = np.asarray(img.tolist(), dtype=np.uint8)
        p = out_dir / f"frame_{i:03d}.png"
        Image.fromarray(arr).save(p)
        paths.append(p)
        print(f"frame {i + 1}/{frames} saved {p}", end="\r")
    print()

    gif = out_dir / "orbit.gif"
    imgs = [Image.open(p) for p in paths]
    imgs[0].save(gif, save_all=True, append_images=imgs[1:], duration=50, loop=0)
    print(f"gif: {gif} ({len(paths)} frames, {gif.stat().st_size // 1024} KB)")


if __name__ == "__main__":
    main()

File: crates/cga-examples/src/bin/demo_engine.rs
// Orbit demo: three.js-style scene + orbit animation -> animated GIF
// (examples/engine/orbit.gif).  Run: cargo run -p cga-examples --bin demo_engine -- [frames]
use cga_core::*;
use cga_examples::{data_f32, mlx_frame_gc};
use cga_gpu::*;
use std::f64::consts::PI;

fn build_scene() -> Scene {
    let mut sc = scene(None);
    // ground plane
    sc.add_mesh(mesh(MeshParams {
        geometry: Geometry::PlaneGeometry(plane_geometry([0.0, 1.0, 0.0], 0.0)),
        material: standard_material(MaterialParams {
            color: color_hex(0xB0B0B0),
            roughness: 0.7,
            metalness: 0.0,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [0.0, 0.0, 0.0],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    // red sphere
    sc.add_mesh(mesh(MeshParams {
        geometry: Geometry::SphereGeometry(sphere_geometry(1.0)),
        material: standard_material(MaterialParams {
            color: color_hex(0xC0392B),
            roughness: 0.25,
            metalness: 0.25,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [0.0, 1.0, 0.0],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    // blue sphere
    sc.add_mesh(mesh(MeshParams {
        geometry: Geometry::SphereGeometry(sphere_geometry(0.6)),
        material: standard_material(MaterialParams {
            color: color_hex(0x2980B9),
            roughness: 0.15,
            metalness: 0.35,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [-2.2, 0.6, 0.5],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    // gold cylinder (infinite)
    sc.add_mesh(mesh(MeshParams {
        geometry: Geometry::CylinderGeometry(cylinder_geometry(0.7, -1.0)),
        material: standard_material(MaterialParams {
            color: color_hex(0xD4AC0D),
            roughness: 0.4,
            metalness: 0.3,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [2.2, 0.7, -0.5],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    // green box
    sc.add_mesh(mesh(MeshParams {
        geometry: Geometry::BoxGeometry(box_geometry(0.9, 0.9, 0.9)),
        material: standard_material(MaterialParams {
            color: color_hex(0x27AE60),
            roughness: 0.6,
            metalness: 0.0,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [0.8, 0.45, 1.8],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    // purple disc (tilted)
    sc.add_mesh(mesh(MeshParams {
        geometry: Geometry::CircleGeometry(circle_geometry(0.9)),
        material: standard_material(MaterialParams {
            color: color_hex(0x8E44AD),
            roughness: 0.3,
            metalness: 0.0,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [-2.4, 2.2, 0.8],
        rotation_axis: [1.0, 0.0, 0.0],
        rotation_angle: -0.4,
        motor: None,
    }));
    // refractive glass sphere
    sc.add_mesh(mesh(MeshParams {
        geometry: Geometry::SphereGeometry(sphere_geometry(0.8)),
        material: standard_material(MaterialParams {
            color: color_hex(0xAAD4FF),
            roughness: 0.05,
            metalness: 0.0,
            emissive: color_hex(0x000000),
            opacity: 0.08,
            ior: 1.5,
            absorption: 0.2,
        }),
        position: [0.4, 1.5, 2.6],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    sc.add_light(directional_light(
        color_hex(0xFFFFFF),
        0.38,
        [0.4, 1.0, 0.35],
    ));
    sc.add_light(point_light(color_hex(0xFFFFFF), 0.7, [0.0, 4.0, 3.5]));
    sc.add_light(directional_light(
        color_hex(0xFFFFFF),
        0.18,
        [0.0, 0.35, 0.9],
    ));
    sc.add_light(ambient_light(color_hex(0xFFFFFF), 0.34));
    sc
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let frames: i32 = if args.len() > 1 {
        args[1].parse().unwrap_or(90)
    } else {
        90
    };
    // 以仓库根目录为 CWD 运行
    let out_dir = "examples/engine";

    let sc = build_scene();
    let mut camera = perspective_camera(
        50.0,
        4.0 / 3.0,
        0.1,
        100.0,
        [0.0, 2.4, 6.2],
        [0.0, 0.8, 0.0],
        [0.0, 1.0, 0.0],
    );
    camera.look_at([0.0, 0.8, 0.0], None);
    let controls = orbit_controls([0.0, 0.8, 0.0], 0.0, 0.42, 6.6);
    let mut r = renderer(360, 270, 2, 3);

    let mut gif_frames: Vec<Vec<u8>> = Vec::new();
    for i in 0..frames {
        let mut ctrl = controls;
        ctrl.azimuth = 2.0 * PI * f64::from(i) / f64::from(frames);
        ctrl.elevation = 0.42 + 0.12 * (4.0 * PI * f64::from(i) / f64::from(frames)).sin();
        ctrl.update(&mut camera);
        let img = r.render(sc.clone(), camera);
        gif_frames.push(f32_rgba_to_u8(&data_f32(&img)));
        drop(img);
        // 每帧收尾：drop 释放死 Metal buffer，再把 MLX 不复用的 cache 还给 OS
        mlx_frame_gc();
        println!("frame {}/{} rendered", i + 1, frames);
    }
    save_gif(&format!("{out_dir}/orbit.gif"), &gif_frames, 360, 270, 3);
    println!("saved {out_dir}/orbit.gif");
}
